pisa/population.py

```diff
@@ -0,0 +1,206 @@
+
+        # Extract url for data for the latest year
+        url = data["files"][0]
+        filehandle, _ = urllib.request.urlretrieve(url)
+
+        return filehandle
+
+
+    @staticmethod
+    def process_population_worldpop(file_path: str, admin_area_boundaries: Polygon | MultiPolygon) -> pd.DataFrame:
+        """Processes the downloaded worldpop data into the required format of a dataframe"""
+        df = Population.raster_to_df(file_path, admin_area_boundaries)
+        return df
+
+
+    @staticmethod
+    def raster_to_df(file_path: str, admin_area_boundaries: Polygon | MultiPolygon) -> pd.DataFrame:
+        """
+        Convert raster file to a dataframe of longitude, latitude
+        and statistical population count
+
+        Function takes the bounds of the raster file in the raster_fpath, draws an evenly spaced sequence of points between
+        the xmin & xmax, and between ymin & ymax, and then generates a grid to cover the complete square area inside the
+        boundaries. The population count for each point in the grid that falls within the given MultiPolygon area
+        (identified by the mask) is extracted from the raster file, and a dataframe with latitude, longitude & population
+        count for each point in the raster is returned.
+        """
+
+        # Convert raster file to dataframe
+        with rasterio.open(file_path) as src:
+            # Create 2D arrays
+            xmin, ymax = np.around(src.xy(0.00, 0.00), 9)
+            xmax, ymin = np.around(src.xy(src.height - 1, src.width - 1), 9)
+            x = np.linspace(xmin, xmax, src.width)
+            y = np.linspace(ymax, ymin, src.height)
+            xs, ys = np.meshgrid(x, y)
+            zs = src.read(1)
+            # Adm area mask
+            mask = Population.get_admarea_mask(admin_area_boundaries, src)
+            xs, ys, zs = xs[mask], ys[mask], zs[mask]
+            data = {
+                "longitude": pd.Series(xs),
+                "latitude": pd.Series(ys),
+                "population": pd.Series(zs),
+            }
+            # Create X,Y,Z DataFrame
+            df = pd.DataFrame(data=data)
+        return df
+
+    @staticmethod
+    def get_admarea_mask(
+        vector_polygon: Polygon | MultiPolygon, raster_layer: rasterio.DatasetReader) -> np.ndarray:
+        """
+        Extract mask from raster for a given MultiPolygon
+
+        Return a boolean mask for the raster layer which is True where the (multi)polygon is located and false for all
+        points outside the given (Multi)Polygon
+        """
+        gtraster, bound = mask(
+        raster_layer, [vector_polygon], all_touched = True, crop = False
+
+        )
+        # Keep only non zero values
+        adm_mask = gtraster[0] > 0
+        return adm_mask
+
+    # if you want to add new data source (e.g. geojson):
+    # create method get_population_gson(), add the datasource in the DataSource class and add a call to it in get_population_data()
```

Comment:
I would add this in the docstring of the module or in the abstract class of the population sources, if you decided to go through that avenue (see other comment).